keps/sig-node/3542-cri-image-pulling-with-progress-notification/README.md

```diff
@@ -0,0 +1,896 @@
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+## Motivation
+
+In High Performance Computing and other specialized enviroments container
+images often get extremely big. Pulling such images even over high speed
+links can take a lot of time, tens of minutes or more. This often results
+in a bad user experience when deploying a workload triggers pulling such
+an image as a side-effect. For a long time the user has no visibility to
+how creating the workload progresses, or if it progresses at all.
+
+Introducing a new CRI ImageService RPC call for pulling images with a
+streaming progress notification response would provide the low-level
+building blocks to improve this situation.
+
+
+
+
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Extend CRI to provide image pulling progress API that can be utilized by client
+tools (crictl pull), or machinery (kubelet pulling the image)
+
+### Non-Goals
+
+- implementation of utilization of proposed interface additions by the
+client-tools and / or k8s services
```

Comment:
is how kubelet reads, stores, and exposes the progress also part of non-goals? coz I guess that would be what the majority of the work is. This KEP will be essentially a no-op without kubelet changes, right?

Reply:
Yes indeed, this is an APi extension without implementation of kubelet part, thus trivial. We can have kubelet implementation included into this KEP, of course, but that indeed might require more technical details and potentially discussion. My thinking was to have  CRI extension in 1.26 and another KEP for 1.27 with kubelet implementation with events posting to the Pod object and probably new kubelet config entry support to have default notification setting there.

Reply:
Alright, the freeze is in place, I'll change the description of KEP and issue to actually include the kubelet implementation and target  this KEP to v1.27 as alpha.

Reply:
Thanks Alexey! I think kubelet-side design will definitely help us design CRI protos.

On the other hand, have you looked at container runtime implementations? For example Containerd has the support for pull progress and reporting # of bytes downloaded in https://github.com/containerd/containerd/pull/6150.